### ISSUE TITLE
Sending tensor to device fails when using uneven width-sharded memory config

### ISSUE THREAD

**Issue by esmalTT** [tenstorrent/tt-metal#22396]
# Summary

This is follow on from [#20726](https://github.com/tenstorrent/tt-metal/issues/20726).

Calling `ttnn.to_device` on a host tensor and using a memory config with uneven shards does not work. This is because the uneven sharding config would require padding to be added to the tensor before sending to device.

The issue can be reproduced with the following script:

```python
import torch
import ttnn
import pytest

from tests.ttnn.utils_for_testing import assert_equal


@pytest.mark.parametrize(
    "input_shape, input_layout, input_shard_grid, input_shard_shape, input_shard_orientation, input_sharding_scheme",
    [
        (
            [1, 1, 96, 1],  # PASSES - UNEVEN HEIGHT SHARDED
            ttnn.ROW_MAJOR_LAYOUT,
            ttnn.CoreRangeSet(
                {
                    ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(1, 0)),
                }
            ),
            (64, 1),
            ttnn.ShardOrientation.ROW_MAJOR,
            ttnn.TensorMemoryLayout.HEIGHT_SHARDED,
        ),
        (
            [1, 1, 1, 96],  # FAILS - UNEVEN WIDTH SHARDED
            ttnn.ROW_MAJOR_LAYOUT,
            ttnn.CoreRangeSet(
                {
                    ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(1, 0)),
                }
            ),
            (1, 64),
            ttnn.ShardOrientation.ROW_MAJOR,
            ttnn.TensorMemoryLayout.WIDTH_SHARDED,
        ),
        (
            [1, 1, 1, 96],  # PASS - EVEN WIDTH SHARDED
            ttnn.ROW_MAJOR_LAYOUT,
            ttnn.CoreRangeSet(
                {
                    ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(2, 0)),
                }
            ),
            (1, 32),
            ttnn.ShardOrientation.ROW_MAJOR,
            ttnn.TensorMemoryLayout.WIDTH_SHARDED,
        ),
    ],
)
def test_tensor_creation(
    device,
    input_shape,
    input_layout,
    input_shard_grid,
    input_shard_shape,
    input_shard_orientation,
    input_sharding_scheme,
):
    shard_spec = ttnn.ShardSpec(input_shard_grid, input_shard_shape, input_shard_orientation)
    memory_config = ttnn.MemoryConfig(input_sharding_scheme, ttnn.BufferType.L1, shard_spec)

    torch_input_tensor = torch.randn(input_shape)

    # First way to create a tensor and move it to device
    print("Creating tensor...")
    input_tensor0 = ttnn.Tensor(
        torch_input_tensor, ttnn.bfloat16, device=device, layout=input_layout, mem_config=memory_config
    )

    # Second way to create a tensor and move it to device
    print("Creating tensor using from_torch...")
    input_tensor1 = ttnn.from_torch(
        torch_input_tensor, device=device, dtype=ttnn.bfloat16, layout=input_layout, memory_config=memory_config
    )
    assert_equal(ttnn.to_torch(input_tensor0), ttnn.to_torch(input_tensor1))

    # Third way to create a tensor and move it to device
    print("Creating tensor using from_torch+to_device")
    input_tensor3 = ttnn.from_torch(torch_input_tensor, device=None, dtype=ttnn.bfloat16, layout=input_layout)
    # This fails for uneven width sharded tensors because of page size mismatch
    input_tensor3 = ttnn.to_device(input_tensor3, device=device, memory_config=memory_config)
    assert_equal(ttnn.to_torch(input_tensor0), ttnn.to_torch(input_tensor3))
```

**Comment by esmalTT** [comment]
The other use case that is affected by this issue is `ttnn.copy_host_to_device_tensor()` which is commonly used for trace and/or multi-command queue models.

For example, we are currently forced to use 6 DRAM cores instead of 7 for the BH version of UNet to avoid creating uneven shard specs:

https://github.com/tenstorrent/tt-metal/blob/15163cf252bfc1efbc27f9fe207e58a3b3af1b28/models/experimental/functional_unet/tests/test_unet_trace.py#L41-L55
